lib/src/Uncertainty/Algorithm/MetaModel/TensorApproximation/TensorApproximationAlgorithm.cxx

```diff
@@ -0,0 +1,647 @@
+      }
+    }
+    if (minimumError < errorThreshold)
+    {
+      LOGINFO(OSS() << "Minimum error=" << minimumError << " smaller than threshold=" << errorThreshold);
+      break;
+    }
+
+  }
+
+  // shrink tensor if smaller optimal rank
+  tensor_[marginalIndex].setRank(optimalRank);
+
+  // reset proxies to expose entire sample
+  for (UnsignedInteger i = 0; i < dimension; ++ i)
+  {
+    proxy_[i].setRowFilter(Indices());
+  }
+
+  greedyRankOne(marginalIndex,
+                transformedInputSample_,
+                outputSample_.getMarginal(marginalIndex),
+                marginalResidual,
+                marginalRelativeError);
+}
+
+
+void TensorApproximationAlgorithm::greedyRankOne (const UnsignedInteger marginalIndex,
+                                                  const NumericalSample & x,
+                                                  const NumericalSample & y,
+                                                  NumericalScalar & marginalResidual,
+                                                  NumericalScalar & marginalRelativeError)
+{
+  const UnsignedInteger r = tensor_[marginalIndex].getRank();
+  NumericalSample yRes(y);
+  for (UnsignedInteger k = 0; k < r; ++ k)
+  {
+    rankOne(marginalIndex,
+            tensor_[marginalIndex].rank1tensors_[k],
+            x,
+            yRes,
+            marginalResidual,
+            marginalRelativeError);
+
+    // build basis
+    NumericalMathFunctionCollection prodColl(k + 1);
+    for (UnsignedInteger k2 = 0; k2 <= k; ++ k2)
+    {
+      Pointer<NumericalMathEvaluationImplementation> p_evaluation = new RankOneTensorEvaluation(tensor_[marginalIndex].rank1tensors_[k2]);
+      prodColl[k2] = NumericalMathFunction(p_evaluation);
+    }
+    const Basis basis(prodColl);
+    Indices indices(prodColl.getSize());
+    indices.fill();
+
+    // update alpha
+    PenalizedLeastSquaresAlgorithm algo(x, y, basis, indices);
+    algo.run();
+    NumericalPoint rk(algo.getCoefficients());
+    for (UnsignedInteger k2 = 0; k2 <= k; ++ k2)
+    {
+      tensor_[marginalIndex].rank1tensors_[k2].radius_ = rk[k2];
+    }
+
+    // compute residual
+    NumericalMathFunction combination(prodColl, rk);
+    yRes = y - combination(x);
+
+  }
+}
+
+void TensorApproximationAlgorithm::rankOne (const UnsignedInteger marginalIndex,
+                                            RankOneTensor & rank1Tensor,
+                                            const NumericalSample & x,
+                                            const NumericalSample & y,
+                                            NumericalScalar & marginalResidual,
+                                            NumericalScalar & marginalRelativeError)
+{
+  Bool convergence = false;
+  const UnsignedInteger dimension = x.getDimension();
+  const UnsignedInteger size = x.getSize();
+  NumericalPoint weight(size, 1.0 / size);
+  UnsignedInteger r = tensor_[marginalIndex].getRank();
+  Collection<Collection<Basis> > subBasis(r);
```

Comment:
A 'Collection<Collection>' may hurt performance, but fortunately this variable is unused :-D